### PR TITLE
ddl: arg v2 for modify column and multi schema change

### PR DESCRIPTION
**Pull request by joechenrh**
### What problem does this PR solve?

Issue Number: ref https://github.com/pingcap/tidb/issues/54436

Problem Summary:

### What changed and how does it work?

Refactor args for Modify Columns.

Since all job arguments are refactored, multi schema change will be refactored in this PR too.

### Check List

Tests 

- [X] Unit test
- [ ] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No need to test
  > - [ ] I checked and no code files have been changed.
  > 

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

Please refer to [Release Notes Language Style Guide](https://pingcap.github.io/tidb-dev-guide/contribute-to-tidb/release-notes-style-guide.html) to write a quality release note.

```release-note
None
```

### CODE REVIEW

**Review by D3Hunter — COMMENTED** [review]
seems this pr need to change, as we have merged the args inside https://github.com/pingcap/tidb/pull/56130, will review after that

File: pkg/ddl/ddl.go [hunk 1]
```diff
@@ -205,6 +205,21 @@ type JobWrapper struct {
 	cacheErr error
 }
 
+// FillArgsWithSubjobs fill args for job and its sub jobs
+func (jobW *JobWrapper) FillArgsWithSubjobs() {
```
Comment (D3Hunter) [inline pkg/ddl/ddl.go:209]: it's quite common convention to put methods of struct after ctor

File: pkg/ddl/job_submitter.go [hunk 2]
```diff
@@ -316,10 +316,9 @@ func (s *JobSubmitter) addBatchDDLJobs2Table(jobWs []*JobWrapper) error {
 	for _, jobW := range jobWs {
 		job := jobW.Job
 		if job.Version == 0 {
-			// if not set, fix it to version 1
-			// TODO replace this with assert after we add code v2 for all jobs.
-			job.Version = model.JobVersion1
+			return errors.Errorf("Job version should not be zero")
```
Comment (D3Hunter) [inline pkg/ddl/job_submitter.go:319]: replace it with assert is enough, here is to confirm that code in ddl/executor.go set Version explicitly

File: pkg/ddl/job_submitter.go [hunk 3]
```diff
@@ -364,9 +363,7 @@ func (s *JobSubmitter) addBatchDDLJobs2Queue(jobWs []*JobWrapper) error {
 
 		for _, jobW := range jobWs {
 			if jobW.Version == 0 {
-				// if not set, fix it to version 1
-				// TODO replace this with assert after we add code v2 for all jobs.
-				jobW.Version = model.JobVersion1
+				return errors.Errorf("Job version should not be zero")
```
Comment (D3Hunter) [inline pkg/ddl/job_submitter.go:366]: doitto

File: pkg/ddl/ddl.go (hunk 1 above)
Comment (D3Hunter) [inline pkg/ddl/ddl.go:209]:
```suggestion
func (jobW *JobWrapper) FillArgsWithSubJobs() {
```

**Review by D3Hunter — COMMENTED** [review]
10/18

File: pkg/ddl/create_table.go [hunk 4]
```diff
@@ -160,6 +160,8 @@ func onCreateTable(jobCtx *jobContext, job *model.Job) (ver int64, _ error) {
 	})
 
 	args, err := model.GetCreateTableArgs(job)
+	jobCtx.jobArgs = args
+
```
[the remark this answers is not in the document]
Comment (joechenrh) [reply pkg/ddl/create_table.go:164]: https://github.com/pingcap/tidb/blob/10647c9d733cc66caadae4362e32721a3a41772c/pkg/ddl/schema_version.go#L244-L255
Just saw this TODO when trying to remove CtxVars for Modify Column. So add this by the way, although it's a bit irrelevant Orz.

[the remark this answers is not in the document]
Comment (D3Hunter) [reply pkg/ddl/create_table.go:164]: get, please move after error checking

File: pkg/ddl/ddl.go [hunk 5]
```diff
@@ -226,13 +226,33 @@ func NewJobWrapperWithArgs(job *model.Job, args model.JobArgs, idAllocated bool)
 	}
 }
 
+// FillArgsWithSubJobs fill args for job and its sub jobs
+func (jobW *JobWrapper) FillArgsWithSubJobs() {
+	if jobW.JobArgs != nil {
```
[the remark this answers is not in the document]
Comment (joechenrh) [reply pkg/ddl/ddl.go:231]: multiSchemaChange jobs don't have `JobArgs`

File: pkg/ddl/modify_column.go [hunk 6]
```diff
@@ -166,19 +163,23 @@ func (w *worker) onModifyColumn(jobCtx *jobContext, job *model.Job) (ver int64,
 				oldTempIdxID := tmpIdx.ID
 				setIdxIDName(tmpIdx, newIdxID, tmpIdx.Name /* unchanged */)
 				SetIdxColNameOffset(tmpIdx.Columns[info.Offset], changingCol)
-				modifyInfo.removedIdxs = append(modifyInfo.removedIdxs, oldTempIdxID)
+				redundantIdxs = append(redundantIdxs, oldTempIdxID)
 			}
+
+			args.RedundantIdxs = redundantIdxs
```
Comment (D3Hunter) [inline pkg/ddl/modify_column.go:169]: set it after the for loop?

**Review by D3Hunter — COMMENTED** [review]
rest lgtm

**Review by D3Hunter — APPROVED** [review]

**Review by wjhuang2016 — APPROVED** [review]

**Comment by D3Hunter** [comment]
panics
```
[2024/10/15 09:57:53.538 +00:00] [ERROR] [misc.go:115] ["panic in the recoverable goroutine"] [label=ddl] [funcInfo=submitLoop] [r="runtime error: invalid memory address or nil pointer dereference"] [stack="github.com/pingcap/tidb/pkg/util.Recover
  pkg/util/misc.go:119
runtime.gopanic
  GOROOT/src/runtime/panic.go:785
runtime.panicmem
  GOROOT/src/runtime/panic.go:262
runtime.sigpanic
  GOROOT/src/runtime/signal_unix.go:900
github.com/pingcap/tidb/pkg/meta/model.(*Job).FillArgs
  pkg/meta/model/job.go:489
github.com/pingcap/tidb/pkg/ddl.(*JobWrapper).FillArgsWithSubJobs
  pkg/ddl/ddl.go:239
github.com/pingcap/tidb/pkg/ddl.insertDDLJobs2Table
  pkg/ddl/job_submitter.go:686
github.com/pingcap/tidb/pkg/ddl.(*JobSubmitter).GenGIDAndInsertJobsWithRetry.func1
  pkg/ddl/job_submitter.go:444
github.com/pingcap/tidb/pkg/ddl.genGIDAndCallWithRetry.func1
  pkg/ddl/job_submitter.go:614
github.com/pingcap/tidb/pkg/ddl.genGIDAndCallWithRetry
  pkg/ddl/job_submitter.go:618
github.com/pingcap/tidb/pkg/ddl.(*JobSubmitter).GenGIDAndInsertJobsWithRetry
  pkg/ddl/job_submitter.go:424
github.com/pingcap/tidb/pkg/ddl.(*JobSubmitter).addBatchDDLJobs2Table
  pkg/ddl/job_submitter.go:351
github.com/pingcap/tidb/pkg/ddl.(*JobSubmitter).addBatchDDLJobs
  pkg/ddl/job_submitter.go:119
github.com/pingcap/tidb/pkg/ddl.(*JobSubmitter).submitLoop
  pkg/ddl/job_submitter.go:93
github.com/pingcap/tidb/pkg/ddl.(*ddl).Start.func1
  pkg/ddl/ddl.go:735
github.com/pingcap/tidb/pkg/util.(*WaitGroupWrapper).Run.func1
  pkg/util/wait_group_wrapper.go:157"] [stack="github.com/pingcap/tidb/pkg/util.Recover
  pkg/util/misc.go:115
runtime.gopanic
  GOROOT/src/runtime/panic.go:785
runtime.panicmem
  GOROOT/src/runtime/panic.go:262
runtime.sigpanic
  GOROOT/src/runtime/signal_unix.go:900
github.com/pingcap/tidb/pkg/meta/model.(*Job).FillArgs
  pkg/meta/model/job.go:489
github.com/pingcap/tidb/pkg/ddl.(*JobWrapper).FillArgsWithSubJobs
  pkg/ddl/ddl.go:239
github.com/pingcap/tidb/pkg/ddl.insertDDLJobs2Table
  pkg/ddl/job_submitter.go:686
github.com/pingcap/tidb/pkg/ddl.(*JobSubmitter).GenGIDAndInsertJobsWithRetry.func1
  pkg/ddl/job_submitter.go:444
github.com/pingcap/tidb/pkg/ddl.genGIDAndCallWithRetry.func1
  pkg/ddl/job_submitter.go:614
github.com/pingcap/tidb/pkg/ddl.genGIDAndCallWithRetry
  pkg/ddl/job_submitter.go:618
github.com/pingcap/tidb/pkg/ddl.(*JobSubmitter).GenGIDAndInsertJobsWithRetry
  pkg/ddl/job_submitter.go:424
github.com/pingcap/tidb/pkg/ddl.(*JobSubmitter).addBatchDDLJobs2Table
  pkg/ddl/job_submitter.go:351
github.com/pingcap/tidb/pkg/ddl.(*JobSubmitter).addBatchDDLJobs
  pkg/ddl/job_submitter.go:119
github.com/pingcap/tidb/pkg/ddl.(*JobSubmitter).submitLoop
  pkg/ddl/job_submitter.go:93
github.com/pingcap/tidb/pkg/ddl.(*ddl).Start.func1
  pkg/ddl/ddl.go:735
github.com/pingcap/tidb/pkg/util.(*WaitGroupWrapper).Run.func1
  pkg/util/wait_group_wrapper.go:157"]

```

**Pull request merged**